Improve `bundle validate` output

kanterov opened the pull request:
## Changes
This combination of changes allows pretty-printing errors happening during the "load" and "init" phases, including their locations.

Move to render code into a separate module dedicated to rendering `diag.Diagnostics` in a human-readable format. This will be used for the `bundle deploy` command.

Preserve the "bundle" value if an error occurs in mutators. Rewrite the go templates to handle the case when the bundle isn't yet loaded if an error occurs during loading, that is possible now.

Improve rendering for errors and warnings:
- don't render empty locations
- render "details" for errors if they exist

Add `root.ErrAlreadyPrinted` indicating that the error was already printed, and the CLI entry point shouldn't print it again.

## Tests
Add tests for output, that are especially handy to detect extra newlines

pietern commented on cmd/bundle/utils/utils.go:
```diff
@@ -20,19 +20,16 @@ func ConfigureBundleWithVariables(cmd *cobra.Command) (*bundle.Bundle, diag.Diag
-		return nil, diags
-	}
 
 	return b, diags
```
`b` may still be nil; I expect `root.MustConfigureBundle` to also return `nil` on error.

kanterov replied:
I added a test in case "b" is nil. We might change few other methods as well if they don't preserve bundle while they should.

pietern reviewed (commented):
This is great, thanks.

Please move the `render` package under the top level `bundle` package.

Does the output change at all for existing warnings?

kanterov commented:
@pietern I've manually checked if there is a diff for warnings. There were a couple of extra new lines, and I've added tests for these cases. I also added tests in case the bundle is nil so we don't crash.

I also moved the package as you suggested. Please take a look again.

pietern commented on bundle/render/render_text_output.go:
```diff
@@ -0,0 +1,175 @@
+{{- if .Location.File }}
+  {{ "in " }}{{ .Location.String | cyan }}
+{{- end }}
+
```
No detail for warning?

kanterov replied:
I can add it for consistency, but it wasn't there before, and there is no use case for it right now. We simply put the whole body into "summary". I will add it.

kanterov replied:
Added detail + unit test for it

pietern commented on cmd/bundle/validate.go:
```diff
@@ -137,19 +34,35 @@ func newValidateCommand() *cobra.Command {
+			if err := diags.Error(); err != nil {
+				return diags.Error()
+			} else {
+				panic("failed to load bundle")
```
Please return `fmt.Errorf` with some description. Always better (and friendlier) than a panic.

kanterov replied:
The reason why I added panic is because it should never happen. If this happens, invariant is violated, and we have a bug in our code, so we should crash immediately. I can change it to `fmt.Errorf` if you think it's more appropriate.

pietern commented on cmd/root/root.go:
```diff
@@ -97,7 +97,7 @@ func Execute(cmd *cobra.Command) {
 	// Run the command
 	cmd, err := cmd.ExecuteContextC(ctx)
-	if err != nil {
+	if err != nil && err != ErrAlreadyPrinted {
```
This should use `errors.Is` so that it also works if (for whatever reason) the error is wrapped.

kanterov commented on bundle/render/render_text_output.go:
```diff
@@ -0,0 +1,175 @@
+// RenderTextOutput renders the diagnostics in a human-readable format.
+//
+// It prints errors and returns root.AlreadyPrintedErr if there are any errors.
+// If there are unexpected errors during rendering, it returns an error different from root.AlreadyPrintedErr.
```
Oh, right. Initially I implemented as described, but then realized it's cleaner not to have this behaviour, and return nil instead.
